### DISCUSSION TITLE
Firmware and others subjects

### DISCUSSION THREAD

**Discussion by mehdichaouch** [UniverseDevel/ULO-Manager#3] [Q&A]
Hello @UniverseDevel,

Thanks this repository 🙏, and all the hard work ⛏️ you have done 👍.
ULO-Controller is the first technical element I found about ULO owl. It's like nobody try to hack it 😭.

I switched on mine yesterday (after 3y in cardboard), it has an old firmware. Do you know if last knowed FW 10.1007 is available somewhere? And if there is possibility to doing it by web browser?

After having it working properly, I would like to connect it to Home Assistant, like that no problem with this [secure issue](https://youtu.be/AfVlNBb7N2I).

Best,

**Comment by UniverseDevel** [comment]
Hi,

latest version that I was able to get my ULO updated to was, 10.1308. Recently I got my hands on second ULO that was running older version (6.x that was the version before they tried to fix the problem with camera access without autnetication and also before they introduced a bigger bug that could give you whole control over ULO without authentication) and I would like to try to get it updated, but it is not that easy.

The update link ULO used to call is no longer live for quite some time and I found it after it was already taken down unfotunately. I want to try to create a backup on the ULO with higher version and then resotre this backup into ULO with lower version, but I am currently moving and my ULO with higher version is somewhere in the box and I am unable to get to it.

Also, the ULO that I just received I would like to take apart and get into its OS if possible, but I am not sure if I am able to do so as I don't know much how or if it is possible to access it. At this moment I only know they used Android as OS.

Until I find my older ULO I cannot verify if update to higher version is possible. Other than that I don't think there are any other oprions to do so. Sorry.

**Reply by mehdichaouch** [reply]
Thank you for all the info.

About the dead link, do you know the url? Maybe there is some stuff in [archive.org](https://archive.org ).

Mine has firmware: 08.0701.
After configuring it, I done a backup in micro sd card. The backup zip, has inside a folder `files\firmware`, I don't know the goal of it but it's maybe useful?

**Comment by UniverseDevel** [comment]
The URL was pointing to AWS, but I did not keep it as it was dead. I might try to get it again when I find some time.

Backup, backs up also the firmware as you mentioned. Thats why I assume that with backup you should be also able to change versions. But I am not able to test this until I find my other ULO.

**Reply by mehdichaouch** [reply]
Ping me here when you'll be back hands in your ULO 😉. Have a nice moving house.

**Reply by SnapperNewbie** [reply]
I tried to get the IP where ULO checks the FW. If i send the POST request "/api/v1/interface/CheckVersionOnCloud" to ULO, my router catches this IP: 34.232.121.46 - points to an amazon server: https://whatismyipaddress.com/ip/34.232.121.46 

IP reverse lookup DNS: ec2-34-232-121-46.compute-1.amazonaws.com

**Reply by UniverseDevel** [reply]
The IP seems to be correct and even trace to Amazon cloud checks out. Unfortunately this endpoint is long dead. I have added this information into [README.md](https://github.com/UniverseDevel/ULO-Controller/blob/master/README.md).

**Comment by SnapperNewbie** [comment]
Hi Guys!

Like mehdi already said, thanks for your effort!
I'm in the same boat with this Ulo thing. I stumbled upon it today, and have no idea if this thing have any usecase. ATM im on FW 08.0807 - Useless... I also thought of an integration to HomeAssistant, but i cant figure out if i can access the video stream. With "MotionEye" i have no luck.

**Reply by UniverseDevel** [reply]
Video steam should be accessible on `ws://<ulo ip>/api/v1/live` and I believe it should be accessible from HomeAssistant. That's the one thing I never had time to test out, but at this moment I don't see a reason why it wouldn't work. I even have suspicion that this WebSocket stream is not secured and can be accessed without authentication (which is bad from security point of view, but good for fiddling around).

**Reply by mehdichaouch** [reply]
@SnapperNewbie in HA (HomeAssistant) maybe you can integrate the link of Ulo static picture.
You can found [here](https://youtu.be/AfVlNBb7N2I) how to get it.
Let me know if you found a solution to integrate Ulo in HA, because I plan to test it.

**Reply by SnapperNewbie** [reply]
@mehdichaouch i had success! After fiddelig around and investigate the `main.js` code from the web gui i found an usable link for the Livestream! the link to the stream: `rtsp://<ulo_ip>:8901/live`

Shocking fact: no password needed :D
My Firmware Version: 08.0807

In HA i installed the HACS Integration "[WebRTC Camera](https://github.com/AlexxIT/WebRTC)" 
For the card configuration i take the "loginPicturce.jpg" while the Stream is loading, and the "rtsp" link:

```
type: custom:webrtc-camera
url: rtsp://<ulo_ip>:8901/live
poster: http://<ulo_ip>/media/loginPicture.jpg
```


![grafik](https://github.com/UniverseDevel/ULO-Controller/assets/35278091/5aec1da8-7f99-43a9-bbb1-cdb4100352d7)

![ulo rtsp stream](https://github.com/UniverseDevel/ULO-Controller/assets/35278091/10e9d09b-b0dc-4e63-992d-10b4068148bd)

**Reply by UniverseDevel** [reply]
Great work! I always wondered if there is any underlying rtsp stream. I will add that to documentation for reference.

**Reply by SnapperNewbie** [reply]
@UniverseDevel hold on! i just recognized, the stream only works if im logged in and watch the stream in an browser. So, if any user is logged in, the stream is available. Maybe i find a solution for that. :/

**Reply by UniverseDevel** [reply]
I have discovered that security hole before and reported it to Mu Design CS (see [SECURITY_WARNING.md](https://github.com/UniverseDevel/ULO-Controller/blob/master/SECURITY_WARNING.md)). From specific version when they fixed that stream image was not accessible before logging in, I noticed that when whoever is logged in, anyone can access ULO without authentication. This is even worse that it works over the internet as well so do not make ULO accessible from internet. It is possible that this security hole was there even before the fix.

**Reply by mehdichaouch** [reply]
@SnapperNewbie big thanks for you tech feedback, you are few step before me 👏 .
About the need of login to access to the stream, have you try to set url with this kind of pattern?
```
url: rtsp://<user>:<password>@<ulo_ip>:8901/live
poster: http://<user>:<password>@<ulo_ip>/media/loginPicture.jpg
```

**Reply by SnapperNewbie** [reply]
Hi @mehdichaouch!

Yes, i tried to access the stream with this pattern, unfortunately it doesn't work. It seems to me that the login process starts this rtsp stream. Have no idea if there could be a workaround for this.

**Comment by UniverseDevel** [comment]
Could you do me a favor and perform following steps:

1. Create a backup of your ULO on SD card
2. Put SD card into PC and find zip file in `config` folder usually: `config\ulo_yyyymmdd_hhmiss.zip`
3. Open this zip and inside of this zip go into folder `files`
4. Extract folder `firmware`
5. Inside of the zip, get out of the folder `files` and open file `ulo.json` in for example text editor
6. Inside of this file check for JSON path `$.config.firmware.currentversion` (usually XX.XXXX format where X are version numbers)
7. Copy value from this JSON path and rename folder `firmware` that you extracted before to value from `$.config.firmware.currentversion`
8. Zip this folder and find a way how to get it to me or share it here [Firmware](https://github.com/UniverseDevel/ULO-Controller/tree/master/Firmware)

Thanks.

**Reply by mehdichaouch** [reply]
Done [here](https://github.com/UniverseDevel/ULO-Controller/pull/4) with firmware 08.0701 😉

**Reply by UniverseDevel** [reply]
Thanks, it's merged.

**Comment by UniverseDevel** [comment]
I got good news and bad news:
- The good news is that I finally found my old ULO
- The bad is that restoring newer firmware to ULO with older version did not work

When I did the restore, ULO UI was telling me that it has a newer version. But it behaved the same way as before and some fixes they introduced were missing. Also, the API was telling me version 6 instead of 10. When I refreshed the UI and cleared the cache as well the version number got rolled back. Not even system log mentiones any attempt to update anything. So, firmware is not so easily updated with restore.

I have some ideas what to try next, but I will need more time to test it.

**Reply by mehdichaouch** [reply]
Hello @UniverseDevel,
I'm back on my Ulo because I have now an Home Assistant!
I plan to use Ulo as a cam detection for house or for my Octoprint when I 3D print...
So did you get time to try your ideas?

**Reply by UniverseDevel** [reply]
Hi, I did try couple of things and none of them worked. After a full week of attempts without success I abandoned ULO all together. I needed a reliable camera and ULO just is not up to the task.
